File: dotnet/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/FoodProduct.cs
namespace VisitorDesignPattern.Example1;

/// <summary>
/// Concrete element: represents a food product.
/// Has an additional property (IsOrganic) specific to this type.
/// </summary>
public class FoodProduct : IProduct
{
    public string Name { get; }
    public decimal Price { get; }
    public bool IsOrganic { get; }

    public FoodProduct(string name, decimal price, bool isOrganic)
    {
        Name = name;
        Price = price;
        IsOrganic = isOrganic;
    }

    public void Accept(IVisitor visitor)
    {
        visitor.Visit(this);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/src/FoodProduct.ts
import { IProduct } from './IProduct';
import { IVisitor } from './IVisitor';

export class FoodProduct implements IProduct {
  readonly name: string;
  readonly price: number;
  readonly isOrganic: boolean;

  constructor(name: string, price: number, isOrganic: boolean) {
    this.name = name;
    this.price = price;
    this.isOrganic = isOrganic;
  }

  accept(visitor: IVisitor): void {
    visitor.visitFood(this);
  }
}
